[FlexibleHeader] Fixed behavior for top layout guide and safe area insets.

jverkoey opened the pull request:
This PR is a rolled-forward version of https://github.com/material-components/material-components-ios/pull/3754. The original PR had to be rolled back while we were cutting v55.1.0 due to the fact that it introduced behavioral changes that would have negatively affected clients.

This roll-forward maintains the entirety of the existing flexible header behavior unless `.topLayoutGuideAdjustmentEnabled = true`.

The primary change between this PR and the previous PR is that all new behavior is now gated by the `topLayoutGuideAdjustmentEnabled` property.

---

# Original PR description

This change introduces more predictable behavior surrounding topLayoutGuide and additionalSafeAreaInsets when using a flexible header (fhvc) container view controller (vc).

The purpose of this change is to support flexible header container view controllers with no-tracking-scroll-view based content.

## Background

Prior to this change, the fhvc was not adjusting the topLayoutGuide of the content vc when using a MDCFlexibleHeaderContainerViewController. The container vc was also not setting the bounds and autoresizing mask of the content view controller, leading to an oddly positioned content vc with no awareness of its occluding header. Many of our catalog examples are implemented with incorrect layout because of this, and I've seen a few examples of client usage that's attempting to work around this behavior.

The content vc's top layout guide is not being adjusted was because the flexible header view was always assuming that its parent vc was the vc whose topLayoutGuide should be updated. This is not the case when using a container vc - the flexible header vc is a sibling to the content vc, with the parent being the container view controller.

## Possible solutions

There were two paths to fixing this that I could see:

1. Make the fhvc a child of the content view controller. This would require zero changes to the fhvc, but may introduce behavioral changes to clients. This would require introducing a somewhat tricky behavioral flag (on change, we'd have to reparent the fhvc).
2. Allow the fhvc to keep updating the container vc's top layout guide, and constrain the container vc's top layout guide to the content vc's top layout guide.
2. Make it possible for the fhvc to target a specific view controller's top layout guide. This will also be a behavioral change, but a somewhat easier one to gate.

In any path, we need to also set the content vc's frame to match the container vc's bounds and set an autoresizing mask accordingly.

This PR represents option 3 because it represents the cleanest, most explicit mechanism for updating the correct top layout guide. We don't care if the container's top layout guide gets updated (option 2), and changing the fhvc hierarchy for containers is a fairly involved task that I feel could be tackled separately.

## This change

After this change, clients can enable a new flag on the container view controller to enable the desired topLayoutGuide behavior.

For example:

```swift
container.topLayoutGuideAdjustmentEnabled = true
```

This will ensure that the topLayoutGuide and additionalSafeAreaInsets of the content view controller are modified when the flexible header's height changes.

This runtime flag introduces new behavior and is off by default to avoid creating a breaking change. We will eventually enable this flag by default once all internal clients have migrated to the new behavior. The breaking behavior is primarily due to the fact that we are now properly setting the content view controller’s frame and autoresizing mask, which can affect client layouts.

This change is the first in a series of changes that will allow us to better support scroll view-less view controllers that otherwise have no way to position their content below the flexible header. Scroll view-less controllers are most commonly wrapped in a container view controller.

## Upcoming deprecations

This change marks `updateTopLayoutGuide` on MDCFlexibleHeaderViewController as to-be-deprecated because it exposes an implementation detail.

## Follow-up work

I will add this same flag and behavior to AppBar's container vc.

I will then enable this behavior in the MDCCatalog on a case-by-case basis so that we can fix layouts by using the top layout guide and/or safe area insets.

---

Closes: https://github.com/material-components/material-components-ios/issues/2471
Closes https://github.com/material-components/material-components-ios/issues/4192

ianegordon commented on components/FlexibleHeader/src/MDCFlexibleHeaderContainerViewController.m:
```diff
@@ -96,6 +101,42 @@ - (void)setContentViewController:(UIViewController *)contentViewController {
+- (void)updateTopLayoutGuideBehavior {
+  if (_topLayoutGuideAdjustmentEnabled) {
+    if ([self isViewLoaded]) {
+      self.contentViewController.view.translatesAutoresizingMaskIntoConstraints = YES;
```
Won't this overwrite the contentViewController state?
What if the client has intentionally set translateAutoresizing to NO, or has a different autoresizingMask?

jverkoey replied:
Yes, that is intentional. As an owner of this view controller it is our responsibility to lay it out, similar to a view owning a subview.

yarneo commented on .kokoro:
```diff
@@ -43,6 +43,7 @@ fix_bazel_imports() {
   rewrite_tests "s/import MaterialComponents.Material(.+)_(.+)/import components_\1_\2/"
   rewrite_tests "s/import MaterialComponents.Material(.+)Scheme/import components_schemes_\1_\1/"
   rewrite_tests "s/import MaterialComponents.MaterialMath/import components_private_Math_Math/"
+  rewrite_tests "s/import MaterialComponents.MaterialUIMetrics/import components_private_UIMetrics_UIMetrics/"
```
should we get rid of all the MaterialUIMetrics in this PR?

jverkoey replied:
Moved to https://github.com/material-components/material-components-ios/pull/4294

yarneo commented on components/FlexibleHeader/src/MDCFlexibleHeaderViewController.m:
```diff
@@ -131,22 +124,16 @@ - (void)didMoveToParentViewController:(UIViewController *)parent {
-
-  // On moving to parentViewController, we calculate the height
-  self.flexibleHeaderViewControllerHeightOffset = [self headerViewControllerHeight];
+  [self updateTopLayoutGuide];
```
should this be under a flag?

jverkoey replied:
Done.

yarneo commented on components/FlexibleHeader/src/MDCFlexibleHeaderViewController.m:
```diff
@@ -131,22 +124,16 @@ - (void)didMoveToParentViewController:(UIViewController *)parent {
 - (void)viewWillAppear:(BOOL)animated {
   [super viewWillAppear:animated];
 
-  for (NSLayoutConstraint *constraint in self.parentViewController.view.constraints) {
```
do we want to remove this even if the flag is false?

jverkoey replied:
Fixed.

jverkoey commented:
I think I need to revisit the fhvc logic gating approach I took. I probably need to introduce a bool on the fhvc that mimics the fhvc container’s bool flag, rather than depending on whether the target view controller is set.

Commit:
[FlexibleHeader] Fixed behavior for top layout guide and safe area insets.

This PR is a rolled-forward version of https://github.com/material-components/material-components-ios/pull/3754. The original PR had to be rolled back as part of v55.1.0 due to the fact that it introduced behavioral changes that were negatively affecting clients.

This roll-forward maintains the existing flexible header behavior unless `container.topLayoutGuideAdjustmentEnabled = true`.

The primary change between this PR and the previous PR is that all new behavior is now gated by the `topLayoutGuideAdjustmentEnabled` property and `fhv_shouldUseIdealTopLayoutGuideBehavior` method.

---

This change introduces more predictable behavior surrounding topLayoutGuide and additionalSafeAreaInsets when using a flexible header (fhvc) container view controller (vc).

The purpose of this change is to support flexible header container view controllers with no-tracking-scroll-view based content.

Prior to this change, the fhvc was not adjusting the topLayoutGuide of the content vc when using a MDCFlexibleHeaderContainerViewController. The container vc was also not setting the bounds and autoresizing mask of the content view controller, leading to an oddly positioned content vc with no awareness of its occluding header. Many of our catalog examples are implemented with incorrect layout because of this, and I've seen a few examples of client usage that's attempting to work around this behavior.

The content vc's top layout guide is not being adjusted was because the flexible header view was always assuming that its parent vc was the vc whose topLayoutGuide should be updated. This is not the case when using a container vc - the flexible header vc is a sibling to the content vc, with the parent being the container view controller.

There were two paths to fixing this that I could see:

1. Make the fhvc a child of the content view controller. This would require zero changes to the fhvc, but may introduce behavioral changes to clients. This would require introducing a somewhat tricky behavioral flag (on change, we'd have to reparent the fhvc).
2. Allow the fhvc to keep updating the container vc's top layout guide, and constrain the container vc's top layout guide to the content vc's top layout guide.
2. Make it possible for the fhvc to target a specific view controller's top layout guide. This will also be a behavioral change, but a somewhat easier one to gate.

In any path, we need to also set the content vc's frame to match the container vc's bounds and set an autoresizing mask accordingly.

This PR represents option 3 because it represents the cleanest, most explicit mechanism for updating the correct top layout guide. We don't care if the container's top layout guide gets updated (option 2), and changing the fhvc hierarchy for containers is a fairly involved task that I feel could be tackled separately.

After this change, clients can enable a new flag on the container view controller to enable the desired topLayoutGuide behavior.

For example:

```swift
container.topLayoutGuideAdjustmentEnabled = true
```

This will ensure that the topLayoutGuide and additionalSafeAreaInsets of the content view controller are modified when the flexible header's height changes.

This runtime flag introduces new behavior and is off by default to avoid creating a breaking change. We will eventually enable this flag by default once all internal clients have migrated to the new behavior. The breaking behavior is primarily due to the fact that we are now properly setting the content view controller’s frame and autoresizing mask, which can affect client layouts.

This change is the first in a series of changes that will allow us to better support scroll view-less view controllers that otherwise have no way to position their content below the flexible header. Scroll view-less controllers are most commonly wrapped in a container view controller.

This change marks `updateTopLayoutGuide` on MDCFlexibleHeaderViewController as to-be-deprecated because it exposes an implementation detail.

I will add this same flag and behavior to AppBar's container vc.

I will then enable this behavior in the MDCCatalog on a case-by-case basis so that we can fix layouts by using the top layout guide and/or safe area insets.

---

Closes: https://github.com/material-components/material-components-ios/issues/2471
Closes https://github.com/material-components/material-components-ios/issues/4192

material-automation commented:
## API diff detected the following changes

### FlexibleHeader

#### MDCFlexibleHeaderContainerViewController

*new* property: `topLayoutGuideAdjustmentEnabled` in `MDCFlexibleHeaderContainerViewController`

#### MDCFlexibleHeaderViewController(ToBeDeprecated)

*new* category: `MDCFlexibleHeaderViewController(ToBeDeprecated)`

*modified* method: `-updateTopLayoutGuide` in `MDCFlexibleHeaderViewController(ToBeDeprecated)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCFlexibleHeaderViewController` |
| To: | `c:objc(cy)MDCFlexibleHeaderViewController@ToBeDeprecated` |

#### MDCFlexibleHeaderViewController

*new* property: `topLayoutGuideAdjustmentEnabled` in `MDCFlexibleHeaderViewController`

*new* property: `topLayoutGuideViewController` in `MDCFlexibleHeaderViewController`

<!-- identifier: apidiff -->

Commit:
Better flag.

jverkoey commented:
This is no longer blocked for review. Will address the CI failures.

Commit:
Replace more legacy behavior.

Commit:
Minor cleanup.

jverkoey commented:
I'd like to aim to get this in to the next release so that we can ensure there are zero regressions over the next week.

Commit:
Add non-legacy injection tests.
